### DISCUSSION TITLE
Table component updates

### DISCUSSION THREAD

**Discussion by jamie-budibase** [Budibase/budibase#13366] [Announcements]
We’ve updated how we treat tables in Budibase. You may have noticed we have a few different ways to make tables with the Table component, Table, Block, Grid Block, etc. It can get a little confusing, so we’ve decided to consolidate.

**What does this practically mean?**
- The **Table Block** and **Grid Block** components have been removed.
- The **Table** component has been updated to work in the same way the **Grid Block** component did previously.
- We've added a new screen template "**Table with Side Panel**". This allows you to quickly create a screen in the old **Table Block** format - creating and editing a row using a side panel.

If you’re using the Table Block or Grid block in your existing applications, then no worries - everything will still work as expected. However, moving forward, you’ll have to use the updated Table component.

**Comment by Zergei64** [comment]
What's been also removed is the Grid component that allowed us to position any components within an established grid. This was a very helpful tool for laying out our components however we wanted. In my case, I was requested to have our forms resemble the paper version of our forms and the Grid layout allowed us to do just that. Without it, how can we position components freely in the form to accomplish this?

An example of what I used to be able to do with the Grid layout was something like this:

![image](https://github.com/Budibase/budibase/assets/22362990/3f995d38-d082-4d9f-8cfc-50aace25c07e)

![image](https://github.com/Budibase/budibase/assets/22362990/4f4287be-76bc-4e47-aadc-c2628f2ecfe9)

**Reply by aptkingston** [reply]
Hey @Zergei64. The grid layout component was never intended to be released - it was something I had been playing around with as an idea, but it definitely wasn't ready for release. It was accidentally shipped in a PR when it shouldn't have been. I'm very glad you found it useful though! We will eventually circle back to it and improve out layout further, but it's not on the immediate radar.

**Comment by Sagestum** [comment]
I really miss the "old" table. The new style is very limited when used with a query. 
With the old table I can hide columns from a query, but with the new one this does not seem to be possible. 
Although the new one does lack the options for Row Count. 

It is really much better when editing directly in a table, but with query the old one was really much better in my opinion and it is very sad that I cannot use the old one anymore. 

Old
![query_old](https://github.com/Budibase/budibase/assets/11883823/7fd70411-08a2-4020-9d31-847f0fee6cf2)
New
![query_new](https://github.com/Budibase/budibase/assets/11883823/778f06f6-25d0-4ed0-82c0-57a504b0c31d)

**Reply by aptkingston** [reply]
You should use your query directly as the data source for the new table / grid block rather than a data provider, unless you have a specific reason you need to use a data provider. The new table is similar to the old table block in that it fetches data itself and does not require a data provider.

While your example here should still work (and this is a bug that it doesn't) it's a very unusual setup. If you change the "data" setting to be your query then I would expect this to work as intended.

**Reply by JCV67** [reply]
I totally agree with you. The table component has a clean look and allows operations such as changing date format (very helpful for non english languages) and others. Uses the space better and allows to control column with.

**Reply by aptkingston** [reply]
@JCV67 the new table component is far better at using space and allowing full control of column width (and height) - it's just all done intuitively by resizing columns rather than having to enter a numeric width. It can also be easily resized by end users which the old table could not.

In terms of date formatting - you can still achieve this by using formula columns to present any data in any format you want. This has the additional benefit of being able to be used in any component rather than only tables.

**Reply by JCV67** [reply]
How do you create search boxes on grid tables?

**Reply by AlfreddGco** [reply]
I can't seem to be able to resize my columns. How is this done?

**Reply by aptkingston** [reply]
You drag the edges to resize them, just like resizing columns works in most other tables. This works anywhere there is a table - both the data section and in apps.

**Reply by AlfreddGco** [reply]
Got it.

Apparently, it can only be done through the published app and it seems like the columns width config won't be saved throughout reloads.

It would be cool that columns' widths are saved locally or something like that or at least that the columns take mínimum width so things can fit in. My id columns take some more space than needed.

But I think it's a good change towards a good direction

**Reply by Sagestum** [reply]
It no longer seems possible to adjust the column width for all settings. It is also apparently no longer possible to store colors. In addition, the formatting of a column can no longer be changed. Furthermore, the index column can no longer be hidden, which I personally find superfluous as it only takes up space.
Unfortunately, I have to say that I liked the old view much better. It's a shame that a solution that worked well has been replaced by one that I find much worse.

**Reply by aptkingston** [reply]
@AlfreddGco the column width config is saved based on how you configure the table or view in the data section. Any changes you make to column order or size are saved. There are some presets above the table in the data section - you might prefer setting them all to "small" for a more compact look. Or just drag to whatever size you want.

When used in apps, users can still reorder or resize columns as they like but this is not saved so they don't affect the settings for other users.

**Reply by aptkingston** [reply]
@Sagestum the index column can be any column you choose - it's whatever data you want to show as the main human-readable information about each row. You can change this to any column you want by choosing "Use as display column" in the column dropdown in the data section. This does not need to be a primary key or ID.
![image](https://github.com/Budibase/budibase/assets/9075550/32c8765e-b643-435a-bf6b-4bc0a7f6f1fb)

**Comment by rcarmstrong** [comment]
Any timeline on when this change will be reflected in the documentation? New to Budibase and have been trying out the "Tutorial: Building a HR portal" on [this doc page](https://docs.budibase.com/docs/mysql-mariadb), and got derailed at the Employees Screen section.

**Comment by 04nd01** [comment]
To me this feels like a major step back. With the old table you could comfortably fit 4 columns on a phone screen and have it look clean and useable.
With the new table, the example in the screenshot already has me losing 80% of horizontal screen space due to the first column being pinned. And then there are three separate scrollbars on the screen.

![screenshot_table](https://github.com/Budibase/budibase/assets/22260436/36c88eed-1a4f-406b-ae2d-daf0dc86fa93)

Unless I am missing something, the only way to have a table of data that can be properly viewed and navigated on mobile is to build it yourself out of cards and a repeater.

**Reply by JCV67** [reply]
Also, I couldn't find a way to create search boxes in grid tables. Grid tables are very closed, less flexible on the way we build screens.
Please bring them back, allow users to choose between grid or classic table component.

**Reply by aptkingston** [reply]
Looks like there's a simple solution we could add of just not pinning the display column on mobile. In terms of column widths, that's completely up to you. You can make them as compact as you like. The only thing you can't do is unpin the first column, which I admit is an oversight for mobile tables. Think that's definitely a change worth making.

**Reply by aptkingston** [reply]
@JCV67 do you mean having a search box actually inside the table? Like inside each row is a search box?

If you mean just adding a search box to search the data in the grid, you can easily do this by either:
- Adding some form fields, and setting up a filter on the table. This is exactly the same as how it worked for the old table too.
- Use the built-in quick searching. Just hover over a column and you'll get a search icon. This isn't supported on all types right now, but it is for easily searchable ones like text and options.
<img width="269" alt="image" src="https://github.com/Budibase/budibase/assets/9075550/ff5d573b-6041-4190-bdbf-6e9806731016">

**Reply by JCV67** [reply]
![Screenshot_20240416_222424_Firefox.jpg](https://github.com/Budibase/budibase/assets/104722906/dd5f7b90-d053-46ce-9c67-a57ee2e38738)

Like this.

**Reply by aptkingston** [reply]
@04nd01 would unpinning the display column solve your issue? That would look like this:

https://github.com/Budibase/budibase/assets/9075550/96e6b1c9-5522-42e0-9c04-ff0a97e92ea7

The gutter is still pinned to allow you to select rows for deletion, as well as add rows if you have that configured. But this would free up a lot more space.

**Reply by 04nd01** [reply]
I think the new table is still quicker to show additional scrollbars instead of just extending to accommodate the content, which doesn't seem like the best option for a mobile phone display.
But yes, not having the display column pinned would free up a lot of space to navigate the table and make it much more usable on smaller displays.

**Reply by aptkingston** [reply]
Maybe that's just preference, but I think scrolling and seeing the full values for each column is much better than seeing a few truncated characters for each column and not having any way to see the full value. It supports touch events to scroll, so you don't need to use any scrollbars - you can just touch and drag like you'd expect. It's very fast.

**Reply by 04nd01** [reply]
I tested a bit more. The table always has a vertical scrollbar that will scroll up and down a few pixels, even if the content is much smaller than the vertical space available. As soon as the width of the content hits around 60% of the available width a horizontal scrollbar will appear. It would make more sense for the scrollbars to appear when they are actually needed i.e. at the moment the content exceeds the size of the table.

**Reply by aptkingston** [reply]
Scrollbars do only appear when needed. The table includes padding at the bottom and right size to account for the size of popovers that are rendered for certain cell types - like options, relationships, long form text etc.

**Comment by JCV67** [comment]
The grid view is more similar to Data Tables than desirable, horizontal scroll bars don´t look nice on apps. Also yes, the index column is unwanted.

One thing I was wrong about. I was able to create search boxes for grid tables:
![imagem](https://github.com/Budibase/budibase/assets/104722906/ce80fef3-dffd-44ca-8dee-b28f0a4f8bea)

**Reply by aptkingston** [reply]
The index column can be any column you choose - it's whatever data you want to show as the main human-readable information about each row. You can change this to any column you want by choosing "Use as display column" in the column dropdown in the data section. This does not need to be a primary key or ID.
![image](https://github.com/Budibase/budibase/assets/9075550/32c8765e-b643-435a-bf6b-4bc0a7f6f1fb)

**Comment by kasgithub8** [comment]
Who thought this update was a good idea?

**Reply by Sagestum** [reply]
From my perception, the application developers seem quite enthusiastic. Personally, I liked it better before this update. The new table view looks pretty ugly overall, and at least to me, it always seems to be poorly integrated into the website, like an embedded iframe.

The old table view, on the other hand, was perfectly integrated into the website.

**Reply by aptkingston** [reply]
What issues are you having? We've continually shipped improvements to the table, and at this stage it's a far better component than the old one.

**Reply by kasgithub8** [reply]
The whole concept of using budibase for us and buying the Enterprise version, was the simplicity of creating apps that were fast and reliable, but nowadays, a perfectly fine component breaks just because of updates that also **reduce** the features of the component

**Reply by kasgithub8** [reply]
> What issues are you having? We've continually shipped improvements to the table, and at this stage it's a far better component than the old one.

@aptkingston just two quick examples that completely break my usage of budibase right now:

1. column sizes, can't set a fixed width anymore, I can drag resize it when the app is published but it just resets, and I don't think that width applies for every app user anyway
2. no custom buttons anymore, I used to have different containers with different looking buttons based on a value of a field but that's just not possible anymore

**Reply by aptkingston** [reply]
Thanks for replying :) To address your points:

1. There's already a PR up for this: https://github.com/Budibase/budibase/pull/13610. As I've mentioned before, the widths will default to whatever size you have them in the data section, but with this PR you can just set a custom width again for each table. So this one will be solved very soon.
2. Tables have a setting for exactly this: 
![image](https://github.com/Budibase/budibase/assets/9075550/3c08f6d2-e7f8-49f6-8c1d-825b189d9280)
![image](https://github.com/Budibase/budibase/assets/9075550/1a3ed250-9582-4134-bc2e-56d74d977f6a)
You can add up to 3 custom buttons which do anything you want to each row.

**Reply by aptkingston** [reply]
@kasgithub8 I'd also just like to point out that we did not break any existing tables. We never change existing apps and this is no exception - all your old tables will remain unchanged. The change only affects adding new tables going forwards. There are many more features in the new table compared to the old one, but yes we did remove very little used features like background colours for entire columns.

**Reply by kasgithub8** [reply]
@aptkingston 
1. Is the custom width for the tables or the columns?
2. Yeah, I know there's buttons, but I had distinctive designs for buttons based on values, which I could do with CSS, but it was just so much easier before, I just added containers with components inside the table and they would be displayed in front of all the other columns in a style which ever I wanted.
3. I'm sorry about that. Tho, I had the old table on multiple screens, and it broke on one of my screens, but the other ones aren't broken. So that's my bad.

So yeah, features I would want to see back would be column sizes adjustable inside the component & custom containers and components inside the table **or** being able to display different buttons on conditions!

Edit: and the buttons in front of the columns not at the end, even though they're sticky

Edit 2: And easier Table CSS back, I can't seem to find the class for every table header now. I used to have a background color in the header.

**Reply by aptkingston** [reply]
It's per column, the same as it was for the old table. You can still drag and resize the columns in the app, but this setting will set their initial width (and override the saved width from the data section).

Thanks for clarifying what features you'd like to see.

**Reply by kasgithub8** [reply]
@aptkingston now my old tables are broken even on pages I didn't touch. Plus, when I upgraded to 2.27.1, the new tables only showed 2 out of 6 columns were displayed in the table. When I rolled the app back to the older version, it worked again.

**Reply by aptkingston** [reply]
@kasgithub8 we have not made any changes to the old tables in recent updates, so it sounds like something else is going on. Could you please raise an issue with more details and screenshots? That will help us look in to your issue.

**Comment by Sagestum** [comment]
> What issues are you having? We've continually shipped improvements to the table, and at this stage it's a far better component than the old one.

Let's compare the two. Up is the old one, down is the new one. Which do you prefer? 

![Bildschirmfoto 2024-05-15 um 09 55 41](https://github.com/Budibase/budibase/assets/11883823/52550629-a35b-443d-bb61-1bb818100a47)

here options for columns: 

Old 
![Bildschirmfoto 2024-05-15 um 09 58 05](https://github.com/Budibase/budibase/assets/11883823/e53328b8-4f45-410f-b991-646f7c78b350)
new 

![Bildschirmfoto 2024-05-15 um 09 57 56](https://github.com/Budibase/budibase/assets/11883823/3f3217fb-5004-4769-a8dc-7246fcfc3c09)


...so at this stage it is much better than the old one? Maybe in some cases... but in my case it is quite inferior to the previous version.

**Reply by aptkingston** [reply]
You're right that there were more settings on the old table, but they really weren't useful (other than width and value, but formula columns are the correct way to transform values into something new). The settings for background color and text color applied to an entire column (rather than being conditional on the value of a cell) which means there are no use cases for them other than wanting a multi coloured table like in your screenshot.

We're adding back a width setting (https://github.com/Budibase/budibase/pull/13610) so you'll be able to customise widths for each table to be something different than your data section widths.

So just to clarify, you'd like the static color settings added back?

**Reply by Sagestum** [reply]
well kind of a yes kind of a no, I really like to add color to tables, but on a condition would be the best possible option, but I don't know if this would be a possible option.

**Reply by JCV67** [reply]
Always start a new app using a copy of an old one, delete everything but one old table that you will be able to copy inside the app. I was forced to do that for a table that I intended to use in a modal form, the result with the new tables was very bad so I started using this way.

**Comment by jmlaurenti** [comment]
Hi @jamie-budibase @aptkingston ,

I also thing the new table is kind of a mixed bag, maybe it was a bit too early to force using it as a full replacement of previous (and quite limited table). I fully understand the extra-cost of maintaining old and new components, and that new components might be much better for maintenance and evolution, but the transition was quite brutal if I may.

On my side, the points below are the one preventing me to replace my existing tables with new ones.

**- Ability to embed components in first column :** 
The buttons that we can add on last column are not as flexible as the ability to embed components. For instance, the "dropdown list "plugin which can be used to create a context menu on a row (by the way, the new table buttons are not displaying any icon !)
        ![image](https://github.com/Budibase/budibase/assets/19287401/9259ad7c-a04f-45de-8036-4193d22a2cc2)

**- Display Column pinned as first column**
This is an arbitrary choice which is fine for the Data panel, but very annoying in normal usage. 
ex: I have readonly display column (named displayname) build for that purpose in my mySQL table (by concatenating first name and last name and other information). As I'm displaying first name and last name columns, and because it is calculated, the display column is irrelevant to be displayed in my editing screens.
![image](https://github.com/Budibase/budibase/assets/19287401/1735b392-9b0b-40c6-b480-4a2fc5af45b2)

**-  No column/cell colors**
Ok, this was quite limited in old table...but better than nothing. Of course, ideal would be to have conditional colors for cells ! :-) 

Thanks for your attention :-)

**Comment by rvt** [comment]
I am new to Budibase... so I might be missing something...
but is this the reason why I cannot make the same look and feel in my new app as compared with the Near-Miss-Register? I was trying to replicate that look and feel because I like it, but I cannot find the right settings to make it look the same, even with teh LINK button on the left..

Update: Indeed I found out because of the Table update

**Comment by dado-ca** [comment]
Buttons inside table component are now floating around. This is so ugly. 


![Buttons inside a table](https://github.com/user-attachments/assets/fa96b6ff-5d6d-4ad6-bc63-ef2ea8225802)

**Reply by aptkingston** [reply]
I'm not sure what you mean by "floating around". They are sticky at the right side of each row so you don't need to scroll to see them. How they look is up to you - you can change the labels and variants if you want.

**Comment by dado-ca** [comment]
Thanks for your answer.

That's what I meant. Maybe it's better to show how it looks in practice. 

The first GIF shows the old table with buttons, while the second one shows the new table with buttons. In the second case, the table is unreadable and cluttered, the rows of the table are not readable at all. While in the first GIF, the table is clearer, the interface is cleaner.

 This is especially visible when the application is accessed via a mobile device. I really think you should revert to the old form of the table.

![Screen_Recording_20240809_210527_Edge_1](https://github.com/user-attachments/assets/f65e2e87-cd87-4cb9-bc32-99a2816108cc)

![Screen_Recording_20240809_210439_Edge_1](https://github.com/user-attachments/assets/ee9f01a0-1532-4bdb-9d4c-456ddd5011d8)

**Reply by aptkingston** [reply]
Thanks for providing those gifs - that helps a lot. That definitely looks like an oversight on our part. We just need to stop making the buttons "sticky" on mobile. We do this same thing for the display column, which is sticky on desktop but not sticky on mobile. Doing the same thing for the button columns looks like a good idea.

**Reply by dado-ca** [reply]
Hi, when will this be solved?

**Comment by AlejandroFNadal** [comment]
My organization uses Budibase, I have left tons of bug reports and posts on community discussions helping people to do things or giving ideas. I am starting with this just to say that I like the tool and most of the stuff that it offers.

We self host. I froze the version in the docker compose at around 2.17 because Budibase was changing too much and changes would periodically break things.

We have recently made a copy of our install and run an update there, to see how BB is looking and if we could move things to version 3.

We see that the tables, which were powerful before, are now quite weak. Things we cannot do anymore:

- Creating text fields inside tables, to be able to edit and save fields directly or run queries from tables on value changes
![image](https://github.com/user-attachments/assets/7cbcb5e0-9eb1-42b2-9dd9-dc0cb3565aa0)

Yes, I know that the new table saves directly to the DB, but that works only if you don't use a custom backend. (we dont expect BB to be able to do that with a custom backend, we also save things through the backend). Also, that feature only makes sense in very simple scenarios. In most scenarios, there are fields which should not be editable like that, or changing one field should change another, all things that can get broken by somebody writing directly in such a table.

- Have links inside tables: This allows to both show text, and also be able to jump to another page.
![image](https://github.com/user-attachments/assets/b886a01d-09b8-403a-8da7-7d553f4b92bb)

It is true that the app is not broken, the old tables still work. But it means that if we want to have tables with the functionalities that we need, we will need to copy paste the old tables, and it feels kinda scary, to depend on a non supported feature anymore.

We see that some of the stuff could be done through the SuperTable component, we are testing. We would like to jump to version 3 because we do see lots of improvements in other parts of Budibase.

Does anyone shares these issues? Do you see yourselves adding back the capacity to have form fields and links inside tables? @aptkingston @joebudi

**Reply by kasgithub8** [reply]
Hi @AlejandroFNadal,

You mentioned a great point with having to rely on a non-supported feature.
The old, customizable tables were great, and so far, we don't see the upside of the “new” table in comparison to the old one.

**Comment by fabriziosolaris** [comment]
Old table was really useful thanks to capability to add icons and image in columns, now it's not possible and, please, don't say to use super table, is really ugly and glitch on large volume of data.
